src/System.Linq/src/System/Linq/ElementAt.cs

```diff
@@ -13,17 +13,25 @@ public static TSource ElementAt<TSource>(this IEnumerable<TSource> source, int i
         {
             if (source == null) throw Error.ArgumentNull("source");
             IPartition<TSource> partition = source as IPartition<TSource>;
-            if (partition != null) return partition.ElementAt(index);
-            IList<TSource> list = source as IList<TSource>;
-            if (list != null) return list[index];
-            if (index >= 0)
+            if (partition != null)
+            {
+                bool found;
+                TSource element = partition.TryGetElementAt(index, out found);
+                if (found) return element;
+            }
+            else
             {
-                using (IEnumerator<TSource> e = source.GetEnumerator())
+                IList<TSource> list = source as IList<TSource>;
+                if (list != null) return list[index];
+                if (index >= 0)
                 {
-                    while (e.MoveNext())
+                    using (IEnumerator<TSource> e = source.GetEnumerator())
                     {
-                        if (index == 0) return e.Current;
-                        index--;
+                        while (e.MoveNext())
+                        {
+                            if (index == 0) return e.Current;
```

Comment:
Unrelated to this PR, I had a random thought:
There are a bunch of places now where we wrap an underlying arbitrary ```IEnumerable<T>``` and are forced to iterate through the wrapper with MoveNext/Current.  I wonder what perf would look like if we added another internal interface that provided a single TryMoveNext that enabled moving forward and getting current all in one operation (either ```bool TryMoveNext(out T current)``` or ```T TryMoveNext(out bool success)``` depending on whether the write barrier makes a measurable difference) so that when we have a chain of these operators, the intermediate ones involve only a single interface dispatch rather than two.  Maybe it wouldn't help or be worth it, but might be something to investigate in the future.